File: csharp/978.longest-turbulent-subarray.cs
/*
 * @lc app=leetcode id=978 lang=csharp
 *
 * [978] Longest Turbulent Subarray
 */

using System;

// @lc code=start
public class Solution {
    public int MaxTurbulenceSize(int[] A) {
        int ret = 1, inc = 1, dec = 1, n = A.Length;
        for (int i = 1; i < n; ++i) {
            if (A[i] > A[i - 1]) {
                inc = dec + 1;
                dec = 1;
                ret = Math.Max(ret, inc);
            } else if (A[i] < A[i - 1]) {
                dec = inc + 1;
                inc = 1;
                ret = Math.Max(ret, dec);
            } else {
                inc = 1;
                dec = 1;
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/978.longest-turbulent-subarray.java
/*
 * @lc app=leetcode id=978 lang=java
 *
 * [978] Longest Turbulent Subarray
 */

// @lc code=start
class Solution {
    public int maxTurbulenceSize(int[] arr) {
        int ret = 1, n = arr.length, p = 0, curr = 1;
        for (int i = 1; i < n; ++i) {
            int c = arr[i] - arr[i - 1];
            if (c == 0) {
                curr = 1;
            } else if (c > 0) {
                if (p < 0) {
                    ++curr;
                } else {
                    curr = 2;
                }
            } else {
                if (p > 0) {
                    ++curr;
                } else {
                    curr = 2;
                }
            }
            p = c;
            ret = Math.max(ret, curr);
        }
        return ret;
    }
}
// @lc code=end

